lib/orogen/gen/tasks.rb

```diff
@@ -300,21 +300,33 @@ def register_for_generation
         # <tt>_time</tt> to be added to the generated class (more specifically,
         # to the +Base+ subclass).
 	module TaskContextGeneration
+
+            # This method generates the relative basepath for generation of all files
+            def basepath
+                s = File.join(namespace.split("::").join(File::SEPARATOR))
+                if !s.empty?
+                    s = s + File::SEPARATOR
+                end
+                s
+            end
+
             # The name of the header file containing the C++ code which defines
             # this task context
             def header_file
-                if external_definition?
-                    library_name, name = self.name.split("::")
-                    File.join("#{library_name.downcase}", "#{name}.hpp")
-                else
-                    File.join("#{project.name.downcase}", "#{basename}.hpp")
-                end
+                File.join(project.name.downcase, basepath, "#{basename}.hpp")
```

Comment:
???? Not good IMO.

This is the gen part. Loading is broken, so you don't have the guarantee that #project will actually return the task context's definition project. I would rather keep the old method until a brave soul refactors gen to use the loaders.